File: qing_voting/generate_res_info_files.m
function generate_res_info_files(set_type, config_file)
try
    eval(config_file);
catch
    keyboard;
end

dir_img = sprintf(Dataset.img_dir, category);
dir_anno = sprintf(Dataset.anno_dir, category);

switch set_type
    case 'train'
        file_cache_VC_data = fullfile(VC.cache_dir, sprintf('%s_%s_train.mat', category, dataset_suffix));
        file_list = sprintf(Dataset.train_list, category);
    case 'test'
        file_cache_VC_data = fullfile(VC.cache_dir, sprintf('%s_%s_test.mat', category, dataset_suffix));
        file_list = sprintf(Dataset.test_list, category);
    otherwise
        error('Error: unknown set_type;\n');
end  

assert(exist(file_cache_VC_data, 'file') > 0);
load(file_cache_VC_data);
assert(exist('r_set', 'var')>0);
assert(exist('feat_set', 'var')>0);

assert(exist(file_list, 'file') > 0);
file_ids = fopen(file_list, 'r');
img_list = textscan(file_ids, '%s %d');
img_num = length(img_list{1});

res_info = cell(1,img_num);
for n=1:img_num
    file_img = sprintf('%s/%s.JPEG', dir_img, img_list{1}{n});
    file_anno = sprintf('%s/%s.mat', dir_anno, img_list{1}{n});
    anno = load(file_anno);
    anno = anno.record;
    
    res_info{n}.img = imread(file_img);
    res_info{n}.layer_feature_dist = r_set{n};
    res_info{n}.layer_feature_ori = feat_set{n};
    res_info{n}.sub_type = anno.objects(img_list{2}(n)).subtype;
    res_info{n}.viewpoint = anno.objects(img_list{2}(n)).viewpoint;
    if mod(n,100) == 0
        disp(n);
    end 
end

VC.res_info = '/media/zzs/4TB/qingliu/qing_intermediate/all_K223_res_info/res_info_%s_%s.mat';
save(sprintf(VC.res_info, category, set_type), 'res_info', '-v7.3');

end % end of function

File: qing_voting_py/generate_res_info_files.py
import math
import pickle
import scipy.io as sio
from scipy.spatial.distance import cdist
from FeatureExtractor import *
from config_voting import *

def generate_res_info_files(category_ls, set_type):
    for category in category_ls:
        dir_img = Dataset['img_dir'].format(category)
        dir_anno = Dataset['anno_dir'].format(category)
        
        file_list = Dataset['{0}_list'.format(set_type)].format(category)
        assert(os.path.isfile(file_list))
        with open(file_list, 'r') as fh:
            content = fh.readlines()
        
        img_list = [x.strip().split() for x in content]
        img_num = len(img_list)
        print('total number of images for {1}: {0}'.format(img_num, category))
        
        # FCD_tmp = '/export/home/qliu24/qing_voting_data/intermediate/feat_alex'
        file_cache_feat = os.path.join(Feat['cache_dir'], '{0}_{1}_{2}_carVC_vMFMM30.pickle'.format(category, dataset_suffix, set_type))
        assert(os.path.isfile(file_cache_feat))
        with open(file_cache_feat, 'rb') as fh:
            _, r_set = pickle.load(fh)
            
        
        layer_feature_dist = [None for nn in range(img_num)]
        # layer_feature_dist_super = [None for nn in range(img_num)]
        sub_type = [None for nn in range(img_num)]
        view_point = [None for nn in range(img_num)]
        for nn in range(img_num):
            layer_feature_dist[nn] = r_set[nn]
            # layer_feature_dist_super[nn] = r_super_set[nn]
            file_anno = os.path.join(dir_anno, '{0}.mat'.format(img_list[nn][0]))
            assert(os.path.isfile(file_anno))
            mat_contents = sio.loadmat(file_anno)
            record = mat_contents['record']
            objects = record['objects']
            
            sub_type[nn] = objects[0,0]['subtype'][0,int(img_list[nn][1])-1][0]
            view_point[nn] = (objects[0,0]['viewpoint'][0,int(img_list[nn][1])-1]['azimuth_coarse'][0,0][0,0], \
                              objects[0,0]['viewpoint'][0,int(img_list[nn][1])-1]['elevation_coarse'][0,0][0,0])
            
            if nn%100 == 0:
                print(nn, end=' ')
                sys.stdout.flush()
            
            
        print('\n')
        
        sfile = VC['res_info'].format(category,set_type+'_carVC_vMFMM30')
        with open(sfile, 'wb') as fh:
            pickle.dump([layer_feature_dist, sub_type, view_point], fh)
            
if __name__=='__main__':
    # objs = ['car','aeroplane','bicycle','bus','motorbike','train']
    objs = ['car']
    generate_res_info_files(objs, 'train')